ruby/ql/lib/codeql/ruby/ast/Control.qll

```diff
@@ -422,6 +429,160 @@ class WhenExpr extends Expr, TWhenExpr {
+ *     puts "Connect with user '#{user}'"
+ *   in connection: {username: } unless username == 'admin'
+ *     puts "Connect with user '#{username}'"
+ *   else
+ *     puts "Unrecognized structure of config"
+ * end
+ * ```
+ */
+class CaseMatch extends ControlExpr, TCaseMatch {
+  private Ruby::CaseMatch g;
+
+  CaseMatch() { this = TCaseMatch(g) }
+
+  final override string getAPrimaryQlClass() { result = "CaseMatch" }
+
+  /**
+   * Gets the expression being matched. For example, `foo` in the following example.
+   * ```rb
+   * case foo
+   * in 0
+   *   puts 'zero'
+   * in 1
+   *   puts 'one'
+   * end
+   * ```
+   */
+  final Expr getValue() { toGenerated(result) = g.getValue() }
+
+  /**
+   * Gets the `n`th branch of this case expression, either an `InClause` or a
+   * `StmtSequence`.
+   */
+  final Expr getBranch(int n) {
+    toGenerated(result) = g.getClauses(n)
+    or
+    n = count(g.getClauses(_)) and toGenerated(result) = g.getElse()
+  }
+
+  /**
+   * Gets a branch of this case expression, either an `InClause` or an
+   * `StmtSequence`.
+   */
+  final Expr getABranch() { result = this.getBranch(_) }
+
+  /** Gets the `n`th `in` clause of this case expression. */
+  final InClause getInClause(int n) { result = this.getBranch(n) }
+
+  /** Gets an `in` clause of this case expression. */
+  final InClause getAnInClause() { result = this.getABranch() }
+
+  /** Gets the `else` branch of this case expression, if any. */
+  final StmtSequence getElseBranch() { result = this.getABranch() }
+
+  /**
+   * Gets the number of branches of this case expression.
+   */
+  final int getNumberOfBranches() { result = count(this.getBranch(_)) }
+
+  final override string toString() { result = "case ... in" }
+
+  override AstNode getAChild(string pred) {
+    result = super.getAChild(pred)
+    or
+    pred = "getValue" and result = this.getValue()
+    or
+    pred = "getBranch" and result = this.getBranch(_)
+    or
+    pred = "getInClause" and result = this.getInClause(_)
+    or
+    pred = "getElseBranch" and result = this.getElseBranch()
+  }
+}
+
+/**
+ * An `in` clause of a `case` expression.
+ * ```rb
+ * case foo
+ * in [ a ] then a
+ * end
+ * ```
+ */
+class InClause extends Expr, TInClause {
+  private Ruby::InClause g;
+
+  InClause() { this = TInClause(g) }
+
+  final override string getAPrimaryQlClass() { result = "InClause" }
+
+  /** Gets the body of this case-in expression. */
+  final Stmt getBody() { toGenerated(result) = g.getBody() }
+
+  /**
+   * Gets the pattern in this case-in expression. In the
+   * following example, the pattern is `Point{ x:, y: }`.
+   * ```rb
+   * case foo
+   * in Point{ x:, y: }
+   *   x + y
+   * end
+   * ```
+   */
+  final CasePattern getPattern() { toGenerated(result) = g.getPattern() }
+
+  /**
+   * Gets the pattern guard condition in this case-in expression. In the
+   * following example, the pattern guard condition is `x > 10`.
```

Comment:
``the pattern guard condition is `x > 10` `` -> ``there are two conditions `x > 10` and `x < 10` ``